feat(systemtags): add setting to block non admin to create system tags

Altahrim opened the pull request:
## Summary
Allows admins to restrict tag creation to admins only

Switch with: 
```
occ conf:app:set systemtags --type boolean only_admins_can_create --value 0|1
```

## TODO
- [x] Add frontend switch in system tags settings
- [x] Change display on tag selection for users who can't create tags
- [x] Block color changing too

## Checklist
- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Altahrim committed:
feat(systemtags): add setting to block non admin to create system tags

Signed-off-by: Benjamin Gaussorgues <benjamin.gaussorgues@nextcloud.com>

nfebe committed:
feat(systemtags): toggle for system tag creation in admin settings

Signed-off-by: nfebe <fenn25.fn@gmail.com>

nfebe commented:
### Frontend switch

| Enabled | Disabled |
| ---------- | ------------  |
| ![enabled-tag-creation-non-admins](https://github.com/user-attachments/assets/068642d6-51eb-40b6-8846-a3cb65164cdf) | ![disabled-tag-creation-non-admins](https://github.com/user-attachments/assets/cddc02d9-405d-40f5-a6c0-6d8f9f512f2e) |


### Error view

| Error | NA |
| --------| ----  |
| ![error-cant-update-systag](https://github.com/user-attachments/assets/bebbf6cb-e298-438b-ab9b-ea0a60a5517c) | NA |

nfebe committed:
feat: Update UI to respect tag creation permissions

Added support for the `only_admin_can_create` flag in system tags.
The UI now hides the option to create tags when this flag is enabled, ensuring compliance with admin settings.

Signed-off-by: nfebe <fenn25.fn@gmail.com>

provokateurin reviewed (approved):
Only minor things, LGTM

Altahrim commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -319,6 +328,22 @@ public function canUserAssignTag(ISystemTag $tag, ?IUser $user): bool {
+			return \OC::$CLI;
+		}
+
+		if ($this->appConfig->getValueBool('systemtags', 'only_admins_can_create', false) === false) {
+			return true;
+		}
+
+		return $this->groupManager->isAdmin($user->getUID());
```
I feel it gives less info: if it's  `everyone_can_create = false` you don't know who can create tags (even if it's probably an admin)

skjnldsv commented on apps/systemtags/src/components/SystemTagsCreationControl.vue:
```diff
@@ -0,0 +1,77 @@
+		</h4>
+
+		<p class="settings-hint">
+			{{ t('settings', 'Allow only admins to create tags (enable or disable).') }}
```
```suggestion
			{{ t('settings', 'Restrict tag creation to admins only (enable or disable).') }}
```

skjnldsv replied:
```suggestion
			{{ t('settings', 'If enabled, regular accounts will be restricted from creating new tags but will still be able to assign and remove them from their files.') }}
```

Weird phrasing otherwise and not proper a11y description of the following NcCheckboxRadioSwitch.
You shouldn't have to specify "enable/disable" if the main description convey the direction (Thus my suggestion: use "restrict")

skjnldsv commented on apps/systemtags/src/components/SystemTagsCreationControl.vue:
```diff
@@ -0,0 +1,77 @@
+		<NcCheckboxRadioSwitch type="switch"
+			:checked.sync="onlyAdminsCanCreateEnabled"
+			@update:checked="updateSystemTagsDefault">
+			{{ t('settings', 'Enable') }}
```
```suggestion
			{{ t('settings', 'Restrict tag creation to admins only') }}
```

skjnldsv replied:
```suggestion
			{{ t('settings', 'Restrict tag creation to admins only') }}
```

nfebe commented:
/compile

nfebe commented:
/compile

Altahrim commented:
/compile

Altahrim commented:
/compile amend

nfebe commented:
/compile

artonge commented on apps/settings/lib/Settings/Admin/Server.php:
```diff
@@ -53,6 +53,9 @@ public function getForm() {
 		$this->initialStateService->provideInitialState('profileEnabledByDefault', $this->isProfileEnabledByDefault($this->config));
 
+		// Basic settings
+		$this->initialStateService->provideInitialState('onlyAdminsCanCreateSystemTags', $this->appConfig->getValueString('systemtags', 'only_admins_can_create', 'true'));
```
```suggestion
		$this->initialStateService->provideInitialState('restrictSystemTagsCreationToAdmin', $this->appConfig->getValueString('systemtags', 'restrict_creation_to_admin', 'true'));
```

Altahrim replied:
Yep, that's better :)

artonge reviewed (commented):
Nitpicking on the naming. But not blocking :)

Altahrim commented:
/compile

nfebe commented:
/compile

nextcloud-command committed:
chore(assets): Recompile assets

Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

skjnldsv commented on apps/settings/lib/Settings/Admin/Server.php:
```diff
@@ -53,6 +54,9 @@ public function getForm() {
 		$this->initialStateService->provideInitialState('profileEnabledByDefault', $this->isProfileEnabledByDefault($this->config));
 
+		// Basic settings
+		$this->initialStateService->provideInitialState('restrictSystemTagsCreationToAdmin', $this->appConfig->getValueString('systemtags', 'restrict_creation_to_admin', 'true'));
```
The value is NOT stored as a boolean in the backend, but as an integer @Altahrim 
![image](https://github.com/user-attachments/assets/0b838394-2f4c-4fdc-8332-a0cf69a3bacb)

skjnldsv commented on apps/systemtags/src/files_actions/bulkSystemTagsAction.ts:
```diff
@@ -34,6 +36,11 @@ export const action = new FileAction({
 	// If the app is disabled, the action is not available anyway
 	enabled(nodes) {
+		// By default, everyone can create system tags
+		if (loadState('settings', 'restrictSystemTagsCreationToAdmin', '0') === '1' && getCurrentUser()?.isAdmin !== true) {
+			return false
+		}
```
@nfebe why are we preventing the entire bulk tagging action if creation is disabled?
It does NOT mean we cannot apply tags....

skjnldsv replied:
Will fix in followup
